### PR TITLE
[New Rule] Google Workspace New OAuth Login from Third-Party Application

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issues
* https://github.com/elastic/security-team/issues/5915

## Summary
This rule detects the first time a third-party application has logged in with credentials to Google Workspace and access was requested/granted via OAuth. This is a new terms rule that looks for the third-party ID within the last 30 days. This is important to differentiate between existing known applications and potentially unwanted/malicious third-party applications where adversaries may have compromised service accounts or credentials.

## Additional Information
* `event.dataset: "google_workspace.token"` - Narrows in on OAuth logs.
* `event.action: "authorize"` - Focuses on authorization only where a user has granted the third-party application access.
* `google_workspace.token.scope.data.scope_name: *Login` - Focuses on logins from that third-party application only which is why initial access was added alongside defense evasion for this MITRE mappings.
* `google_workspace.token.client.id: *apps.googleusercontent.com` - Limits to custom third-party applications such as mobile device private apps, desktop apps, etc.

Note that the following new terms was added to limit this only to third-party applications that have never been seen before in the organization's Google Workspace ecosystem within the last 30 days.

```sql
[rule.new_terms]
field = "new_terms_fields"
value = ["google_workspace.token.client.id"]

[[rule.new_terms.history_window_start]]
field = "history_window_start"
value = "now-30d"
```

Screenshot 2023-03-30 at 12 53 33 PM

<details>
<summary> Example data </summary>

```sql
{
  "_index": ".ds-logs-google_workspace.token-default-2023.03.28-000001",
  "_id": "0nv63i6xzD6AGHq1yD5Fq145pyI=",
  "_version": 1,
  "_score": 0,
  "_source": {
    "agent": {
      "name": "ubuntu-server-tdejesus",
      "id": "f4011165-a1b8-4d8e-b902-56ba8835cc28",
      "ephemeral_id": "4c4fa27f-35e7-465b-8383-0b6e6cea1884",
      "type": "filebeat",
      "version": "8.6.2"
    },
    "elastic_agent": {
      "id": "f4011165-a1b8-4d8e-b902-56ba8835cc28",
      "version": "8.6.2",
      "snapshot": false
    },
    "source": {
      "ip": "2603:6010:ed00:b789:1481:7d06:9275:be9c",
      "user": {
        "domain": "dejesusarcheology.com",
        "name": "terrance",
        "id": "115903088752625509360",
        "email": "user"
      }
    },
    "tags": [
      "forwarded",
      "google_workspace-token"
    ],
    "cloud": {
      "availability_zone": "us-east1-b",
      "instance": {
        "name": "ubuntu-server-tdejesus",
        "id": "1709224677170316971"
      },
      "provider": "gcp",
      "machine": {
        "type": "e2-medium"
      },
      "service": {
        "name": "GCE"
      },
      "project": {
        "id": "elastic-security-dev"
      },
      "account": {
        "id": "elastic-security-dev"
      }
    },
    "input": {
      "type": "httpjson"
    },
    "@timestamp": "2023-03-30T16:02:51.432Z",
    "ecs": {
      "version": "8.6.0"
    },
    "related": {
      "hosts": [
        "dejesusarcheology.com"
      ],
      "ip": [
        "2603:6010:ed00:b789:1481:7d06:9275:be9c"
      ],
      "user": [
        "115903088752625509360",
        "terrance",
        "terrance@dejesusarcheology.com"
      ]
    },
    "google_workspace": {
      "kind": "admin#reports#activity",
      "etag": "\"1rpbeKGlFXXrXfE6I4jSKiwRKeqWfVrVpA_lfqSodas/dlZDQwNVHtkn0cYy8DnaSfn5KAA\"",
      "token": {
        "app_name": "Google Chrome",
        "scope": {
          "data": [
            {
              "product_bucket": [
                "IDENTITY"
              ],
              "scope_name": "https://www.google.com/accounts/OAuthLogin"
            }
          ],
          "value": [
            "https://www.google.com/accounts/OAuthLogin"
          ]
        },
        "client": {
          "id": "77185425430.apps.googleusercontent.com",
          "type": "NATIVE_DESKTOP"
        }
      }
    },
    "data_stream": {
      "namespace": "default",
      "type": "logs",
      "dataset": "google_workspace.token"
    },
    "organization": {
      "id": "C00qtspd5"
    },
    "event": {
      "agent_id_status": "verified",
      "ingested": "2023-03-30T16:12:50Z",
      "provider": "token",
      "created": "2023-03-30T16:12:49.659Z",
      "kind": [
        "event"
      ],
      "action": "authorize",
      "id": "8833551346588831865",
      "dataset": "google_workspace.token"
    },
    "user": {
      "domain": "dejesusarcheology.com",
      "name": "terrance",
      "id": "115903088752625509360",
      "email": "terrance@dejesusarcheology.com"
    }
  },
  "fields": {
    "elastic_agent.version": [
      "8.6.2"
    ],
    "source.user.email": [
      "user"
    ],
    "cloud.availability_zone": [
      "us-east1-b"
    ],
    "google_workspace.token.scope.data.scope_name": [
      "https://www.google.com/accounts/OAuthLogin"
    ],
    "source.user.name.text": [
      "terrance"
    ],
    "source.ip": [
      "2603:6010:ed00:b789:1481:7d06:9275:be9c"
    ],
    "agent.name": [
      "ubuntu-server-tdejesus"
    ],
    "event.agent_id_status": [
      "verified"
    ],
    "event.kind": [
      "event"
    ],
    "user.id": [
      "115903088752625509360"
    ],
    "cloud.instance.name.text": [
      "ubuntu-server-tdejesus"
    ],
    "input.type": [
      "httpjson"
    ],
    "data_stream.type": [
      "logs"
    ],
    "related.user": [
      "115903088752625509360",
      "terrance",
      "terrance@dejesusarcheology.com"
    ],
    "google_workspace.token.client.type": [
      "NATIVE_DESKTOP"
    ],
    "tags": [
      "forwarded",
      "google_workspace-token"
    ],
    "cloud.machine.type": [
      "e2-medium"
    ],
    "cloud.provider": [
      "gcp"
    ],
    "event.provider": [
      "token"
    ],
    "agent.id": [
      "f4011165-a1b8-4d8e-b902-56ba8835cc28"
    ],
    "cloud.service.name": [
      "GCE"
    ],
    "ecs.version": [
      "8.6.0"
    ],
    "event.created": [
      "2023-03-30T16:12:49.659Z"
    ],
    "google_workspace.token.app_name": [
      "Google Chrome"
    ],
    "organization.id": [
      "C00qtspd5"
    ],
    "agent.version": [
      "8.6.2"
    ],
    "related.hosts": [
      "dejesusarcheology.com"
    ],
    "source.user.name": [
      "terrance"
    ],
    "user.name": [
      "terrance"
    ],
    "cloud.instance.id": [
      "1709224677170316971"
    ],
    "agent.type": [
      "filebeat"
    ],
    "event.module": [
      "google_workspace"
    ],
    "user.email": [
      "user"
    ],
    "related.ip": [
      "2603:6010:ed00:b789:1481:7d06:9275:be9c"
    ],
    "elastic_agent.snapshot": [
      false
    ],
    "source.user.id": [
      "115903088752625509360"
    ],
    "user.domain": [
      "dejesusarcheology.com"
    ],
    "google_workspace.token.client.id": [
      "77185425430.apps.googleusercontent.com"
    ],
    "elastic_agent.id": [
      "f4011165-a1b8-4d8e-b902-56ba8835cc28"
    ],
    "google_workspace.token.scope.value": [
      "https://www.google.com/accounts/OAuthLogin"
    ],
    "data_stream.namespace": [
      "default"
    ],
    "google_workspace.kind": [
      "admin#reports#activity"
    ],
    "event.action": [
      "authorize"
    ],
    "event.ingested": [
      "2023-03-30T16:12:50.000Z"
    ],
    "@timestamp": [
      "2023-03-30T16:02:51.432Z"
    ],
    "google_workspace.etag": [
      "\"1rpbeKGlFXXrXfE6I4jSKiwRKeqWfVrVpA_lfqSodas/dlZDQwNVHtkn0cYy8DnaSfn5KAA\""
    ],
    "cloud.account.id": [
      "elastic-security-dev"
    ],
    "data_stream.dataset": [
      "google_workspace.token"
    ],
    "agent.ephemeral_id": [
      "4c4fa27f-35e7-465b-8383-0b6e6cea1884"
    ],
    "source.user.domain": [
      "dejesusarcheology.com"
    ],
    "event.id": [
      "8833551346588831865"
    ],
    "event.dataset": [
      "google_workspace.token"
    ],
    "google_workspace.token.scope.data.product_bucket": [
      "IDENTITY"
    ],
    "cloud.instance.name": [
      "ubuntu-server-tdejesus"
    ],
    "cloud.project.id": [
      "elastic-security-dev"
    ],
    "user.name.text": [
      "terrance"
    ]
  }
}
```

</details>

### CODE REVIEW

**Comment by terrancedejesus** [comment]
Not this rule will fail schema validation until https://github.com/elastic/detection-rules/pull/2677 is merged as it updates the integration manifests and schemas.

**Review by imays11 — APPROVED** [review]
Looks good!

File: rules/integrations/google_workspace/defense_evasion_google_workspace_new_oauth_login_from_third_party_application.toml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    Developers may leverage third-party applications for legitimate purposes in Google Workspace such as for administrative tasks.
+    """,
+]
+from = "now-130m"
+index = ["filebeat-*", "logs-google_workspace*"]
+interval = "10m"
+language = "kuery"
+license = "Elastic License v2"
+name = "Google Workspace New OAuth Login from Third-Party Application"
```
Comment (terrancedejesus) [inline rules/integrations/google_workspace/defense_evasion_google_workspace_new_oauth_login_from_third_party_application.toml:24]:
```suggestion
name = "First Time Seen Google Workspace OAuth Login from Third-Party Application"
```

File: rules/integrations/google_workspace/defense_evasion_google_workspace_new_oauth_login_from_third_party_application.toml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+name = "Initial Access"
+reference = "https://attack.mitre.org/tactics/TA0001/"
+
+[rule.new_terms]
+field = "new_terms_fields"
+value = ["google_workspace.token.client.id"]
+
+[[rule.new_terms.history_window_start]]
+field = "history_window_start"
+value = "now-30d"
```
Comment (terrancedejesus) [inline rules/integrations/google_workspace/defense_evasion_google_workspace_new_oauth_login_from_third_party_application.toml:95]:
```suggestion
value = "now-15d"
```

**Review by DefSecSentinel — APPROVED** [review]
LGTM

**Pull request merged**